### PR TITLE
3.0.0 gas optimizations

### PR DESCRIPTION
**Pull request by tbrent**
Two main changes: 
1. RevenueTraderP1.manageToken() now only refreshes if the assets being traded were not refreshed in the current block
2. StRSR.withdraw() now tracks how much RSR it has let out without a refresh, up to 0.5%

Had to do a few things to get StRSR contract size under limit. 

Some of the other stuff in this PR:
- IAssetRegistry was organized wrong, and missing comments
- Adds `lastSave()` to IAsset interface. It's okay that live RTokens will be on the old RTokenAsset missing the function, because the RevenueTrader never calls it on the RTokenAsset
- Made some internal variables private. They aren't used by StRSRVotes so they shouldn't be visible there
- The `Staked` event was missing from cancelUnstake(); fixed

### CODE REVIEW

File: contracts/p1/StRSR.sol [hunk 1]
```diff
@@ -329,13 +334,19 @@ abstract contract StRSRP1 is Initializable, ComponentP1, IStRSR, EIP712Upgradeab
...
 
+        // == Interactions ==
+        leakyRefresh(rsrAmount);
+        IERC20Upgradeable(address(rsr)).safeTransfer(account, rsrAmount);
         emit UnstakingCompleted(firstId, endId, draftEra, account, rsrAmount);
 
-        // == Interaction ==
-        IERC20Upgradeable(address(rsr)).safeTransfer(account, rsrAmount);
+        // == Checks ==
+        require(basketHandler.fullyCollateralized(), "RToken uncollateralized");
```
Comment (julianmrodri) [inline contracts/p1/StRSR.sol:343]: Why moved this checks to the end? They dont make sense before the refresh right?

File: contracts/p1/StRSR.sol [hunk 2]
```diff
@@ -81,8 +81,8 @@ abstract contract StRSRP1 is Initializable, ComponentP1, IStRSR, EIP712Upgradeab
     // draftEra => ({account} => {drafts})
     mapping(uint256 => mapping(address => CumulativeDraft[])) public draftQueues; // {drafts}
     mapping(uint256 => mapping(address => uint256)) public firstRemainingDraft; // draft index
-    uint256 internal totalDrafts; // Total of all drafts {qDrafts}
-    uint256 internal draftRSR; // Amount of RSR backing all drafts {qRSR}
+    uint256 private totalDrafts; // Total of all drafts {qDrafts}
+    uint256 private draftRSR; // Amount of RSR backing all drafts {qRSR}
     uint192 public draftRate; // The exchange rate between drafts and RSR. D18{qDrafts/qRSR}
```
[the remark this answers is not in the document]
Comment (tbrent) [reply contracts/p1/StRSR.sol:86]: Literally just the contract size. exchangeRate() is its inverse, so it doesn't need to be public.

File: contracts/p1/StRSR.sol (hunk 1 above)
Comment (tbrent) [reply contracts/p1/StRSR.sol:346]: they're after the refresh, right?

File: contracts/p1/StRSR.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbrent) [reply contracts/p1/StRSR.sol:86]: Oh sorry I misunderstood. Agree.

File: contracts/p1/StRSR.sol [hunk 3]
```diff
@@ -873,24 +913,30 @@ abstract contract StRSRP1 is Initializable, ComponentP1, IStRSR, EIP712Upgradeab
...
+    // solhint-disable-next-line no-empty-blocks
+    function governanceOnly() private governance {}
+
     /**
      * @dev This empty reserved space is put in place to allow future versions to add new
      * variables without shifting down storage in the inheritance chain.
      * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
      */
-    uint256[30] private __gap;
+    uint256[29] private __gap;
```
[the remark this answers is not in the document]
Comment (tbrent) [reply contracts/p1/StRSR.sol]: Now it's 2 (added gov param var), but yes.

File: contracts/p1/StRSR.sol (hunk 1 above)
Comment (julianmrodri) [reply contracts/p1/StRSR.sol:331]: Yes, was just asking if that was the reason they were moved to the end. I believe they are.

**Pull request merged**